### PR TITLE
support loss function for Self-play Preference Optimization

### PR DESCRIPTION
**Pull request by winglian**
arxiv: https://arxiv.org/abs/2405.00675

Screenshot 2024-05-02 at 8 35 03 AM

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this great addition ! 
Can you add a section here in the docs to mention this method: https://github.com/huggingface/trl/blob/main/docs/source/dpo_trainer.mdx#loss-functions ! 🙏

**Review by younesbelkada — APPROVED** [review]
Thanks a lot !

**Comment by younesbelkada** [comment]
cc @kashif wdyt? 🙏

File: docs/source/dpo_trainer.mdx [hunk 1]
```diff
@@ -111,6 +111,8 @@ The [KTO](https://arxiv.org/abs/2402.01306) authors directly maximize the utilit
 
 The [BCO](https://arxiv.org/abs/2404.04656) authors train a binary classifier whose logit serves as a reward so that the classifier maps {prompt, chosen completion} pairs to 1 and {prompt, rejected completion} pairs to 0. The `DPOTrainer` can be switched to this loss via the `loss_type="bco_pair"` argument.
 
+The [SPPO](https://arxiv.org/abs/2405.00675) authors claim that SPPO is capable of solving the Nash equilibrium iteratively by pushing a to be as large as 1/2 and b to be as small as -1/2 and can alleviate data sparsity issues.
```
Comment (kashif) [inline docs/source/dpo_trainer.mdx:114]:
```suggestion
The [SPPO](https://arxiv.org/abs/2405.00675) authors claim that SPPO is capable of solving the Nash equilibrium iteratively by pushing the chosen rewards to be as large as 1/2 and the rejected rewards to be as small as -1/2 and can alleviate data sparsity issues.
```

**Review by kashif — APPROVED** [review]

**Comment by kashif** [comment]
@winglian  do you want to add an option to the test? e.g.  a `["gpt2", "sppo", True],`

**Review by kashif — APPROVED** [review]

**Comment by angelahzyuan** [comment]
@winglian Thanks for adding our work! @younesbelkada @kashif Just submitted a new pull request at [https://github.com/huggingface/trl/pull/1615](https://github.com/huggingface/trl/pull/1615). This updates the loss function according to Equation (4.8), with $P(y_w > y_l) = 1$ and $P(y_l > y_w) = 0$, and justified it in doc as the hard label version of the algorithm. 

Screenshot 2024-05-02 at 9 01 54 PM


It should work well now for the first iteration. Our reported 3 iterations results was based on the soft label version.

**Comment by flozi00** [comment]
I just gave it a try and it's working better than orpo for me now.
Just installed the main branch this evening with the follow up patch for the hard loss type.
